### PR TITLE
Added more functionalities to the DQ_VrepInterface.

### PR DESCRIPTION
**Pull request by ffasilva**
Mostly related to synchronous simulation and joint torque actuation.

### CODE REVIEW

**Comment by bvadorno** [comment]
Hi @mmmarinho ,

May you take this one, please? 

Cheers,
Bruno

**Comment by mmmarinho** [comment]
@frederico-afonso 
Thanks for your contribution.

1. Could you adjust the synchronous mode API to follow the CPP API? 
https://github.com/dqrobotics/cpp-interface-vrep/blob/master/include/dqrobotics/interfaces/vrep/DQ_VrepInterface.h#L94
```
void set_synchronous(const bool& flag);
void trigger_next_simulation_step();
int wait_for_simulation_step_to_end();
```
    
2. Could you provide a minimal working example including the scene so that I can test it on my side. This can be added to `matlab-examples` later.

@bvadorno I'm asking for the minimal example in case we have to maintain the code if/when/whenever @frederico-afonso is unavailable.

Thanks again,
Murilo

**Comment by frederico-afonso** [comment]
Hi @mmmarinho,

You're right; the code shouldn't be dependent on a specific developer. As soon as I made the requested modifications, I'll send them your way.

Regards,
Frederico

**Comment by frederico-afonso** [comment]
Hi @mmmarinho,

I have added all the requested modifications and a new robot `JacoVrepRobot` for the examples. All the simulations are on the temporary folder `matlab-examples` inside the folder `interfaces->vrep`.

Regards,
Frederico

**Comment by mmmarinho** [comment]
Hello, @frederico-afonso 

Thanks again for the contribution.

Please correct me if I'm wrong:
1. From my first read through, it seems that the following new methods are covered by the minimal examples:

```matlab
set_syncronous()
trigger_next_simulation_step()
wait_for_simulation_step_to_end()
set_joint_torques()
get_joint_torques()
```

But the following might not be covered yet:
```matlab
get_force_sensor_readings()
call_script_function()
get_joint_velocities() %This method is used in JacoVrepRobot.m but not used in the examples.
```
Is my understanding correct? If so, we need all new methods to be somehow covered by the minimal examples.

2. The following method might be missing (not sure if it was supposed to be added or not)
```matlab
set_joint_velocities()
```

3. Given that `JacoVrepRobot` will mostly be used as a reference for future similar classes, we have to consider the naming of the new methods.

Currently, for joint positions we use:
```matlab
get_q_from_vrep()
send_q_to_vrep()
```

For velocity, maybe `qd` can be confused with desired joint position, so maybe
```matlab
get_q_dot_from_vrep()
send_q_dot_to_vrep()
```

For torque information, I'm not sure. 
As suggested by @frederico-afonso 
```matlab
get_tau_from_vrep()
send_tau_to_vrep()
```

@bvadorno, what do you say?

**Comment by frederico-afonso** [comment]
Hello @mmmarinho,

I am happy to contribute.

Answering your questions:

1. You were right. The missing examples were dependent on other local modifications I have made to the DQ Robotics---and are outside the scope of this pull request. I realized that before sending them but forgot to add their replacements to the Github folders.
The new examples are:
`force_sensor_reading_on_vrep` - For `get_force_sensor_readings()` and `call_script_function()`;
`jaco_kinematic_control_on_vrep` - For `get_joint_velocities()`.

2. I never needed to send joint velocities to V-REP; therefore, I have not thought about implementing such a method. For the sake of symmetry, I tried to do it now but did not manage to make it work.
I tried to use both `simxSetObjectFloatParameter` and `simxSetJointTargetVelocity`, but the robot did not move during the simulations. I left this unfinished method commented in the `DQ_VrepInterface` file since it might help future debugs.

3. I agree with you; it might get confusing. I have changed the name of the method as suggested.
As for the methods of get/set torques, I considered using `get/send_torques` but `get/send_tau` felt more appropriate since we already have `get/send_q` instead of `get/send_position`.

Everything should be working now but let me know if you find any more issues.

**Comment by bvadorno** [comment]
Hi @mmmarinho and @frederico-afonso,

I think that for the V-Rep interface we should use 

`get_robot_configuration_from_vrep()` instead of `send_q_to_vrep()`
`set_robot_configuration_from_vrep()` instead of `send_q_to_vrep()` and

`get_configuration_velocities_from_vrep()` and `set_configuration_velocity_from_vrep()` and

`set_torques()` and `get_torques()` (or `get_robot_torques/set_robot_torques`)

The reason for that is that, although in the DQ Robotics community we use `q` for configuration and `tau` for torques, we should prevent any confusion with third-party nomenclature. This way, it's better to be explicit and not adopt a particular convention for variable names.

Also, we should avoid the term "joints", unless we want to refer exclusively to fixed-base manipulators. Otherwise, it's best to say "configuration", which is more general and mathematically appropriate. However, just "configuration" can lead to confusion because "configuration" may have other meanings in the context of a computer program. Hence, my suggestion for using "robot_configuration".

On the other hand, just `configuration_velocities` seems fine because there is no confusion with other meanings for the word "configuration".

@mmmarinho, please let me know what you think.

Cheers,
Bruno

File: interfaces/vrep/matlab-examples/jaco_kinematic_control_on_vrep.m [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+%
+%     DQ Robotics is distributed in the hope that it will be useful,
+%     but WITHOUT ANY WARRANTY; without even the implied warranty of
+%     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+%     GNU Lesser General Public License for more details.
+%
+%     You should have received a copy of the GNU Lesser General Public License
+%     along with DQ Robotics.  If not, see <http://www.gnu.org/licenses/>.
+%
+% DQ Robotics website: dqrobotics.sourceforge.net
```
Comment (mmmarinho) [inline interfaces/vrep/matlab-examples/jaco_kinematic_control_on_vrep.m:24]: Reminder for myself to fix this and other headers.

**Review by mmmarinho — APPROVED** [review]
Thanks again, @frederico-afonso.

@bvadorno I think it is very reasonable. 
Given that those are methods of a class such as `KukaVrep`, if we make an instance called `kuka_vrep`, the following seems proper:
```
kuka_vrep.get_{robot}_configuration()
kuka_vrep.set_{robot}_configuration()
kuka_vrep.get_{robot}_velocity()
%%kuka_vrep.set_{robot}_velocity() <-- This one actually isn't implemented yet.
kuka_vrep.get_{robot}_torque()
kuka_vrep.set_{robot}_torque() 
```
We just need to decide two things:
1- If we add `robot`, which might feel redundant but just `get_configuration()` might be confused with getting some configuration files or parameters.
2- If the noun in the right is singular or plural, as it should probably be the same for all. `configuration/velocity/torque` for example.

Alternatively:
`get_qs_{positions/velocities/torques}` or `get_configuration_space_{positions/velocities/torques}`.

I approved @frederico-afonso's changes and I'll handle the rest myself after the merge.

**Comment by mmmarinho** [comment]
@juanjqo Please check this PR when implementing your changes.

**Comment by bvadorno** [comment]
@mmmarinho and @juanjqo, 

What's the status of this PR? May I merge it with the master branch?

Best wishes,
Bruno

**Comment by bvadorno** [comment]
@juanjqo, do you know if those modifications are compatible with the new ones you intend to add? Please liaise with @ffasilva to ensure that we upload all the contributions in a consistent way.

Thanks,
Bruno

**Comment by juanjqo** [comment]
> @juanjqo, do you know if those modifications are compatible with the new ones you intend to add? Please liaise with @ffasilva to ensure that we upload all the contributions in a consistent way.
> 
> Thanks, Bruno

Hi @bvadorno, no, they are not compatible. At least, some methods are not.  Sure! I'm going to liaise with @ffasilva!

The modifications I am preparing to dqrobotics/matlab aims to synchronize the version of Matlab and C++/Python. The current C++/Python version does not implement methods that return several values or tuples.  For instance, instead of the public method (as proposed in #67)

```matlab
[return_code,output_ints,output_floats,output_trings,retBuffer]=call_script_function(obj,obj_name,script_type,function_name,...
                                                                                    input_ints,input_floats,input_strings,input_buffer,opmode)
````
, which could be used to compute the dynamic properties of a robot, we have specific methods like 

```cpp
    MatrixXd get_inertia_matrix(const std::string& link_name, const REFERENCE_FRAMES& reference_frame=BODY_FRAME, const std::string& function_name = "get_inertia", const std::string& obj_name= "DQRoboticsApiCommandServer");
    MatrixXd get_inertia_matrix(const int& handle, const REFERENCE_FRAMES& reference_frame=BODY_FRAME, const std::string& function_name = "get_inertia", const std::string& obj_name= "DQRoboticsApiCommandServer");

    DQ      get_center_of_mass(const std::string& link_name, const REFERENCE_FRAMES& reference_frame=BODY_FRAME, const std::string& function_name = "get_center_of_mass", const std::string& obj_name= "DQRoboticsApiCommandServer");
    DQ      get_center_of_mass(const int& handle, const REFERENCE_FRAMES& reference_frame=BODY_FRAME, const std::string& function_name = "get_center_of_mass", const std::string& obj_name= "DQRoboticsApiCommandServer");

    double get_mass(const std::string& link_name, const std::string& function_name = "get_mass", const std::string& obj_name= "DQRoboticsApiCommandServer");
    double get_mass(const int& handle, const std::string& function_name = "get_mass", const std::string& obj_name= "DQRoboticsApiCommandServer");

```
that require the script [`DQRoboticsApiCommandServer.lua`](https://github.com/dqrobotics/cpp-examples/blob/master/cmake/vrep_interface_tests/DQRoboticsApiCommandServer.lua). Example of use [here](https://github.com/dqrobotics/cpp-examples/blob/master/cmake/vrep_interface_tests/extracting_inertial_parameters/extracting_inertial_parameters.cpp). The main reason for that is to provide high-level methods for the user (at expense of limited Lua script functionalities).

Furthermore, some methods like `get_joint_torques` and `get_joint_velocities` (and other `get` methods that were not proposed in #67) have different signatures wrt their C++/Python version. For instance:

Matlab:

```matlab
function [joint_torques,return_code]=get_joint_torques()  % Two components are returned
function [joint_velocities,return_code]=get_joint_velocities() % Two components are returned
```
C++:

```cpp
VectorXd get_joint_torques();  
VectorXd get_joint_velocities();  
```

Best regards, 

Juancho

**Comment by mmmarinho** [comment]
@bvadorno @juanjqo 

The way we have to go about this issue (if I remember correctly, please correct me I'm wrong)
1. Close this pull request (the one proposed by @juanjqo supersedes this one)
- https://github.com/dqrobotics/cpp-interface-vrep/pull/1
2. @juanjqo proposes the new pull request in MATLAB.
3. It gets reviewed by @bvadorno (and me?)  
4. Rejoice.

**Comment by bvadorno** [comment]
Agreed. Following this [discussion](https://github.com/dqrobotics/matlab/pull/67#issuecomment-1471572246), I'm closing this issue, then.

@juanjqo, may you start a new pull request with those functionalities in place?

Kind regards,
Bruno

**Comment by juanjqo** [comment]
> Agreed. Following this [discussion](https://github.com/dqrobotics/matlab/pull/67#issuecomment-1471572246), I'm closing this issue, then.
> 
> @juanjqo, may you start a new pull request with those functionalities in place?
> 
> Kind regards, Bruno

@bvadorno sure!

**Pull request closed without merging**